### DISCUSSION TITLE
Generate client from schema

### DISCUSSION THREAD

**Discussion by huehnerlady** [ExpediaGroup/graphql-kotlin#1196] [Q&A] [answered]
Hi,

We are using spring-mvc and therefore cannot use the graphql-kotlin-server dependency. What we did instead is to use the lib `com.expediagroup:graphql-kotlin-schema-generator` to generate our schema, and create the GraphQL-endpoint manually. This works really well and we are very happy with the lib overall ❤️ .

Now we would like to create some contracttests to ensure that our model etc does not break for the client. 
We would like to use graphql-kotlin for it as we use it in other services and like the easy usage. BUT now we have the problem, that we do not have a file an endpoint or anything in that regard to generate the client code in gradle.

Is there maybe another way to generate the client?

**Comment by dariuszkuc** [comment, marked as answer]
Hello 👋 
You should be able to generate your schema file using `graphqlGenerateSDL` task (https://expediagroup.github.io/graphql-kotlin/docs/plugins/gradle-plugin-usage#generating-sdl-example) which you can then use for generating your client.

**Reply by huehnerlady** [reply]
Thank your for your hint.

I tried this but now I seem to have the Problem, that we use interfaces for our beans and therefore have multiple implementations with different spring-conditionals attached to it. When I then attach the `Mutation` interface to our interface, I get the gradle error

```
 > A failure occurred while executing com.expediagroup.graphql.plugin.gradle.actions.GenerateSDLAction
      > Conflicting field names in schema generation [type : Mutation(class: class de.foo.graphql.MutationImpl), field name: update]
```

<deials>
<summary> part of the debug log </summary>

```
2021-06-24T10:57:21.891+0200 [DEBUG] [org.gradle.api.Task] submitting work item to generate SDL for the supported packages = [de.europace.kreditsmart.api.vorgang, de.foo.api.graphql, de.foo.graphql]
2021-06-24T10:57:21.986+0200 [WARN] [generateSDL] No SchemaGeneratorHooksProvider were found, defaulting to NoopSchemaGeneratorHooks
2021-06-24T10:57:22.412+0200 [DEBUG] [generateSDL] Generating schema using hooks = NoopSchemaGeneratorHooks
2021-06-24T10:57:22.412+0200 [DEBUG] [generateSDL]   query classes = [class de.foo.graphql.FooQuery]
2021-06-24T10:57:22.412+0200 [DEBUG] [generateSDL]   mutation classes = [class de.foo.graphql.Mutation, class de.foo.graphql.MutationImpl, class de.foo.graphql.MutationMock]
2021-06-24T10:57:22.412+0200 [DEBUG] [generateSDL]   subscription classes = []
```
</details>


Another question would be how we would use this file in our spring-server. 
If we use a generated SDL file for the clients I would argue with the pattern 'eat your own dogfood' we would also want to use it in our code as well.

Currently we are having the following setup:
```kotlin
  @Bean
  fun graphql(fooQuery: FooQuery, mutation: Mutation): GraphQL {
    val schema = toSchema(
        config = SchemaGeneratorConfig(
            supportedPackages = listOf("de.foo.api.vorgang", "de.foo.api.graphql")
        ),
        queries = listOf(TopLevelObject(fooQuery)),
        mutations = listOf(TopLevelObject(mutation))
    )
    log.info(schema.print())
    return GraphQL.newGraphQL(schema).build()
  }
```

How would we achieve this using the sdl file?

**Reply by dariuszkuc** [reply]
re: conditional queries/mutations - SDL generation is unaware of anything Spring specific and has to be deterministic. You will need to create [custom hooks provider](https://expediagroup.github.io/graphql-kotlin/docs/plugins/hooks-provider) that would know how to generate your schema (i.e. which queries/mutations to pick -> depending on your setup this might be somewhat complex)

re: dogfooding - you can use the generated SDL to generate your [test client](https://expediagroup.github.io/graphql-kotlin/docs/plugins/gradle-plugin-tasks#graphqlgeneratetestclient) that you can then use in your integration tests

**Reply by huehnerlady** [reply]
Yeah we do it the way now to not use the SDL in the service just for generating the client, not perfect, but works for us, thanks for your help :)
